Add support for aggregations over sorted inputs to StreamingAggregation

duanmeng opened the pull request:
Extended AggregationFuzzer to cover new functionality.

Fixes #7665

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -61,8 +61,25 @@ void StreamingAggregation::initialize() {
   aggregates_ = toAggregateInfo(
       *aggregationNode_, *operatorCtx_, numKeys, expressionEvaluator, true);
 
+  // Setup SortedAggregations.
```
nit: perhaps, extract this logic into a helper method

duanmeng replied:
Fixed, added AggregationInfo::extractMaskChannels, SortedAggregation::extractSortedAggregations method, used by both GroupingSet and StreamingAggregation.

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -97,6 +118,17 @@ void StreamingAggregation::initialize() {
         rows_->rowSizeOffset());
   }
 
+  if (sortedAggregations_) {
```
perhaps, extract all logic for initializing aggregate functions into a helper method

duanmeng replied:
Done

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -171,6 +203,11 @@ RowVectorPtr StreamingAggregation::createOutput(size_t numGroups) {
     }
   }
 
+  if (sortedAggregations_) {
```
Shouldn't we skip sorted aggs in the loop above?

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -276,14 +313,23 @@ RowVectorPtr StreamingAggregation::getOutput() {
 
   evaluateAggregates();
 
+  // Initialize new groups and evaluate aggregates in sortedAggregations.
```
Shouldn't we skip sorted aggs in the loop above and in evaluateAggregates?

duanmeng replied:
Fixed

mbasmanova replied:
Somehow I don't see logic to skip sorted aggs in the loop above. See https://github.com/facebookincubator/velox/pull/7764 for context.

duanmeng replied:
Fixed, and move the sortedAggrgation eval into the evaluateAggregates method.

mbasmanova reviewed (commented):
@duanmeng Thank you for adding this capability. Would you extend the fuzzer to cover the new functionality? This should help test it a bit more thouroughly.

duanmeng commented:
@mbasmanova Hi Masha, I've fixed all the comments and added some reusable methods for GroupingSet and StreamingAggregation, could you please help to take a look? I want to put some common logic into AggregateUtils but I am afraid to encounter the circle dependence issue like last time :). 

>Would you extend the fuzzer to cover the new functionality? This should help test it a bit more thoroughly.
Sure, could I extend this in a follow-up?

mbasmanova commented:
> Sure, could I extend this in a follow-up?

I think we need to extend this before landing this PR to make sure we catch as many bugs as possible. Some of the comments I made point out bugs that were not caught by original tests.

duanmeng commented:
> > Sure, could I extend this in a follow-up?
> 
> I think we need to extend this before landing this PR to make sure we catch as many bugs as possible. Some of the comments I made point out bugs that were not caught by original tests.

@mbasmanova OK, got it, will extend it.

duanmeng commented:
>extend the fuzzer to cover the new functionality

@mbasmanova Hi Masha, should I refer this PR https://github.com/facebookincubator/velox/pull/7620 to add a method, say `AggregationFuzzer::verifySortedStreamingAggregation`?

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzer.cpp:
```diff
@@ -1617,6 +1617,20 @@ bool AggregationFuzzer::verifySortedAggregation(
+    if (!groupingKeys.empty()) {
+      plans.push_back(
+          {PlanBuilder()
+               .tableScan(inputRowType)
```
Let's also add a simple plan, but with values node instead of tableScan.

duanmeng replied:
Got it, added. 
I've debugged and run both Velox and Spark aggregation fuzzer. They all reach the breakpoint in the places that I added the streaming aggregation plan, and all success.
```
velox_aggregation_fuzzer_test --seed 0 --duration_sec 300 --enable_sorted_aggregations --logtostderr=1 --minloglevel=0 --repro_persist_path=./tmp/aggregate_fuzzer_repro

spark_aggregation_fuzzer_test --seed 0 --duration_sec 300 --enable_sorted_aggregations --logtostderr=1 --minloglevel=0 --repro_persist_path=./tmp/spark_aggregate_fuzzer_repro
```

mbasmanova replied:
This is great. Would you also run these for longer. 1h each. Just to be absolutely sure.

mbasmanova replied:
Also, please, remove --seed 0.

duanmeng replied:
Got it, will run it nightly :)

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzer.cpp:
```diff
@@ -1617,6 +1617,33 @@ bool AggregationFuzzer::verifySortedAggregation(
              .planNode(),
          splits});
 
+    if (!groupingKeys.empty()) {
```
This should be outside of the isTableScanSupported 'if'.

duanmeng replied:
Fixed

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -239,11 +226,19 @@ void StreamingAggregation::evaluateAggregates() {
   }
+
+  if (sortedAggregations_) {
+    if (!newGroups.empty()) {
```
Let's move this logic out of this method and put it next to function->initializeNewGroups

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -306,4 +304,54 @@ RowVectorPtr StreamingAggregation::getOutput() {
   return output;
 }
 
+std::unique_ptr<RowContainer> StreamingAggregation::createRows(
```
createRows name is a bit confusing; perhaps, makeRowContainer?

mbasmanova reviewed (commented):
@duanmeng Looks great % few small comments. Please, update PR description to accurately describe the changes. It looks outdated.

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzer.cpp:
```diff
@@ -1604,19 +1604,48 @@ bool AggregationFuzzer::verifySortedAggregation(
+    }
     // Set customVerification to false to trigger direct result comparison.
     // TODO Figure out how to enable custom verify(), but not compare().
     testPlans(plans, false, {}, resultOrError, 1);
```
Can you make sure we cll testPlans() even if table scan is not supported?

duanmeng replied:
Got it, I should notice this, moved it out of the if.

duanmeng commented on velox/exec/AggregateInfo.h:
```diff
@@ -79,4 +79,7 @@ std::vector<AggregateInfo> toAggregateInfo(
     std::shared_ptr<core::ExpressionEvaluator>& expressionEvaluator,
     bool isStreaming = false);
 
+std::vector<std::optional<column_index_t>> extractMaskChannels(
```
Fixed

duanmeng reviewed (commented):
@mbasmanova I've resolved all the comments and I will rerun the tow fuzzer test for 1h, could you help to take another look? Thanks a lot.

duanmeng commented on velox/exec/SortedAggregations.cpp:
```diff
@@ -409,4 +409,31 @@ void SortedAggregations::extractValues(
 
+std::unique_ptr<SortedAggregations>
+SortedAggregations::extractSortedAggregations(
+    std::vector<AggregateInfo>& aggregates,
```
Added, and made some code changes in SortedAggregations to make it can be `const &`.

mbasmanova commented on velox/exec/SortedAggregations.h:
```diff
@@ -31,7 +31,12 @@ class SortedAggregations {
+      const RowTypePtr& inputType,
+      memory::MemoryPool* pool);
+
+  static std::unique_ptr<SortedAggregations> create(
```
Please, add comments to explain that this method may return null.

duanmeng replied:
Added.

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -275,14 +266,24 @@ RowVectorPtr StreamingAggregation::getOutput() {
 
-  evaluateAggregates();
+  if (sortedAggregations_) {
+    if (!newGroups.empty()) {
```
Should we put this check around the loop over aggregates_ as well?

duanmeng replied:
Added, and I put all the new group init into a new function initializeNewGroups.

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -221,11 +202,17 @@ const SelectivityVector& StreamingAggregation::getSelectivityVector(
 
-void StreamingAggregation::evaluateAggregates() {
+void StreamingAggregation::evaluateAggregates(
+    const std::vector<vector_size_t>& newGroups) {
```
This parameter is no longer used. Let's remove.

duanmeng replied:
Done, sorry that I forgot to remove it.

mbasmanova commented on velox/exec/StreamingAggregation.cpp:
```diff
@@ -306,4 +284,78 @@ RowVectorPtr StreamingAggregation::getOutput() {
+      continue;
+    }
+
+    if (!newGroups.empty()) {
```
perhaps, move this check to the start of the method and return early

```
if (numGroups_ == numPrevGroups) {
   return;
}
```

mbasmanova commented on velox/exec/tests/utils/AggregationFuzzer.cpp:
```diff
@@ -1604,24 +1604,61 @@ bool AggregationFuzzer::verifySortedAggregation(
   }
 
+  std::vector<PlanWithSplits> plans;
+  plans.push_back(
```
This seems redundant.

mbasmanova reviewed (commented):
Looks good % one nit.

duanmeng commented:
@mbasmanova Hi Masha, the Spark Fuzzer ran 3600s successfully. I've resolved the comments and ran the Velox fuzzer for another 3600s, and it succeeded.

### Velox Fuzzer
```
I20231215 11:03:14.950178 13459763 AggregationFuzzer.cpp:419] ==============================> Done with iteration 480
I20231215 11:03:14.950223 13459763 AggregationFuzzer.cpp:1006] Total functions tested: 42
I20231215 11:03:14.950228 13459763 AggregationFuzzer.cpp:1007] Total masked aggregations: 71 (14.76%)
I20231215 11:03:14.950241 13459763 AggregationFuzzer.cpp:1009] Total global aggregations: 35 (7.28%)
I20231215 11:03:14.950246 13459763 AggregationFuzzer.cpp:1011] Total group-by aggregations: 358 (74.43%)
I20231215 11:03:14.950250 13459763 AggregationFuzzer.cpp:1013] Total distinct aggregations: 48 (9.98%)
I20231215 11:03:14.950254 13459763 AggregationFuzzer.cpp:1015] Total aggregations over sorted inputs: 45 (9.36%)
I20231215 11:03:14.950258 13459763 AggregationFuzzer.cpp:1017] Total window expressions: 40 (8.32%)
I20231215 11:03:14.950263 13459763 AggregationFuzzer.cpp:1019] Total aggregations verified against reference DB: 59 (12.27%)
I20231215 11:03:14.950268 13459763 AggregationFuzzer.cpp:1021] Total aggregations not verified (non-deterministic function / not supported by reference DB / reference DB failed): 234 (48.65%) / 128 (26.61%) / 26 (5.41%)
I20231215 11:03:14.950289 13459763 AggregationFuzzer.cpp:1026] Total failed aggregations: 41 (8.52%)
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

### Spark Fuzzer
```
I20231214 23:51:54.230937 13060886 AggregationFuzzer.cpp:976] ==============================> Done with iteration 614
I20231214 23:51:54.230957 13060886 AggregationFuzzer.cpp:1751] Total functions tested: 9
I20231214 23:51:54.230962 13060886 AggregationFuzzer.cpp:1752] Total masked aggregations: 96 (15.61%)
I20231214 23:51:54.230976 13060886 AggregationFuzzer.cpp:1754] Total global aggregations: 39 (6.34%)
I20231214 23:51:54.230983 13060886 AggregationFuzzer.cpp:1756] Total group-by aggregations: 449 (73.01%)
I20231214 23:51:54.230988 13060886 AggregationFuzzer.cpp:1758] Total distinct aggregations: 73 (11.87%)
I20231214 23:51:54.230991 13060886 AggregationFuzzer.cpp:1760] Total aggregations over sorted inputs: 104 (16.91%)
I20231214 23:51:54.230996 13060886 AggregationFuzzer.cpp:1762] Total window expressions: 54 (8.78%)
I20231214 23:51:54.231001 13060886 AggregationFuzzer.cpp:1764] Total aggregations verified against reference DB: 159 (25.85%)
I20231214 23:51:54.231006 13060886 AggregationFuzzer.cpp:1766] Total aggregations not verified (non-deterministic function / not supported by reference DB / reference DB failed): 179 (29.11%) / 241 (39.19%) / 36 (5.85%)
I20231214 23:51:54.231014 13060886 AggregationFuzzer.cpp:1771] Total failed aggregations: 0 (0.00%)
[==========] Running 0 tests from 0 test suites.
[==========] 0 tests from 0 test suites ran. (0 ms total)
[  PASSED  ] 0 tests.
```

mbasmanova reviewed (approved):
@duanmeng Looks great. Thank you.
